AzureStorageBlob/Source/StorageBlobClient.swift

```diff
@@ -256,47 +300,52 @@ public class StorageBlobClient: PipelineClient, PagedCollectionDelegate {
-
-        // Construct parameters
-        var queryParams = [String: String]()
-
-        // Construct headers
-        var headerParams = HttpHeaders()
-        headerParams[HttpHeader.apiVersion] = self.options.apiVersion
-        headerParams["x-ms-range"] = "bytes=0-33554431"
-        // if let requestId = requestId { headerParams["x-ms-client-request-id"] = requestId }
-
-        // Construct and send request
-        let request = self.request(method: HttpMethod.GET,
-                                   url: url,
-                                   queryParams: queryParams,
-                                   headerParams: headerParams)
-        let context: [String: AnyObject] = [
-            ContextKey.allowedStatusCodes.rawValue: [200, 206] as AnyObject,
-        ]
-        run(request: request, context: context, completion: { result, httpResponse in
+    /**
+     Download a blob from a specific container.
+     - Parameter blob: The name of the blob.
+     - Parameter fromContainer: The name of the container.
+     - Parameter withOptions: A `DownloadBlobOptions` object to control the download operation.
+     - Parameter then: An `HttpResultHandler` closure that returns a `StorageStreamDownloader` object on success.
+     */
+    public func download(blob: String, fromContainer container: String, withOptions options: DownloadBlobOptions? = nil,
+                         then completion: @escaping HttpResultHandler<StorageStreamDownloader>) throws {
+        let downloader = try StorageStreamDownloader(client: self, name: blob, container: container, options: options)
+        downloader.initialRequest() { result, httpResponse in
             switch result {
-            case let .success(data):
-                guard let data = data else {
-                    let noDataError = HttpResponseError.decode("Response data expected but not found.")
-                    completion(.failure(noDataError), httpResponse)
-                    return
-                }
-                completion(.success(data), httpResponse)
+            case .success(_):
+                completion(.success(downloader), httpResponse)
             case let .failure(error):
                 completion(.failure(error), httpResponse)
             }
-        })
+        }
+    }
+
+    /**
+     Download a blob from a specific container.
+     - Parameter url: A URL to a blob to download.
+     - Parameter options: A `DownloadBlobOptions` object to control the download operation.
+     - Parameter completion: An `HttpResultHandler` closure that returns a `StorageStreamDownloader` object on success.
+     */
+    public func download(url: URL, withOptions options: DownloadBlobOptions? = nil,
+                         then completion: @escaping HttpResultHandler<StorageStreamDownloader>) throws {
+        let (host, container, blob) = try parse(url: url)
+        // ensure that if the base URL is altered temporarily because of the URL that it is switched back.
```

Comment:
Could maybe use more description in this comment. From what I have gleaned:

* `baseUrl` is a field (comes from the parent class)
* When calling `download(url:)` we temporarily change the `baseUrl` field and then change it back after the call is complete

If that's correct, I have two comments. One, and this is purely a style thing, but I was confused about where `baseUrl` came from until I realized it was probably a field on the parent class. How do you feel about prefixing fields with `self.`? Two, it seems like this means that each call to `download(url:)` on this client changes the internal state of the client. I feel like this would have to cause problems, for example if I have a client initialized for one storage account, then call `download(url:)` to download a file from a different account (host). While that operation is in flight, it seems like calls to other client methods will go to the wrong place?

Reply:
Swift best practices are that you don't use self unless it is necessary (i.e. in a closure or when a local variable has the same name and you want to target the property). Swiftlint will whine incessantly if you do this.

You raise a great point about messing with the baseUrl. Instead, I changed this to actually instantiate a new client in the event that the baseUrl doesn't match the URL. The scenario here is when you want to download a public blob URL. I added a TODO to reconsider this implementation and testing the scenario should be informative.